paimon-core/src/main/java/org/apache/paimon/utils/FormatReaderMapping.java

```diff
@@ -174,7 +174,7 @@ public FormatReaderMapping build(
                 String formatIdentifier, TableSchema tableSchema, TableSchema dataSchema) {
 
             // extract the whole data fields in logic.
-            List<DataField> allDataFields = fieldsExtractor.apply(dataSchema);
+            List<DataField> allDataFields = new ArrayList<>(fieldsExtractor.apply(dataSchema));
```

Comment:
The fieldsExtractor here directly retrieves the fields list from the `dataSchema`. Adding fields directly can introduce unknown bugs.

So I changed the fields in the `TableSchema` to be unmodifiable.
`this.fields = Collections.unmodifiableList(new ArrayList<>(fields));`